### PR TITLE
Adding Javascript rendering feature

### PR DESCRIPTION
**Pull request by TBonnin**
## What does this change?
Adding Javascript rendering.
Nashorn javascript engine is being used. 
Rendering is being executed within an Akka actor (a single one for now but we can envisage running a pool of actor to improve throughput)

A very simple endpoint has been added to the archive app (`/dev/ui`). We will add it to all apps with parameter support later. This endpoint is not being routed by the router, so it is not accessible by readers

A component is a simple class than define a `def asJson: Option[JsValue]` function.

cc @guardian/dotcom-platform

### CODE REVIEW

**Comment by gtrufitt** [comment]
giphy 4

File: archive/conf/routes [hunk 1]
```diff
@@ -5,5 +5,7 @@
 
 GET        /_healthcheck                            controllers.HealthCheck.healthCheck()
 
+GET        /dev/ui                                  controllers.DevComponentController.renderComponent()
```
Comment (TBonnin) [inline archive/conf/routes]: @sndrs @SiAdcock Does this url work for you?

Comment (sndrs) [inline archive/conf/routes]: yep - great

Comment (SiAdcock) [inline archive/conf/routes]: ❤️  it

**Comment by sndrs** [comment]
giphy-downsized-large

File: common/test/uiComponent/core/RenderingActorTest.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  lazy val actorSystem: ActorSystem = app.actorSystem
+  implicit lazy val timeout = new Timeout(2.seconds)
+
+  class TestRenderingActor extends RenderingActor {
+    override def javascriptFile: String = "components/TestButtonComponent.js"
+  }
+
+  lazy val actor = actorSystem.actorOf(Props(new TestRenderingActor))
+
+  "Sending rendering message" should "return correct string" in {
```
Comment (NataliaLKB) [inline common/test/uiComponent/core/RenderingActorTest.scala]: Misleading name of test. Should be changed to `Sending rendering message should return a string`

File: project/Prototypes.scala [hunk 3]
```diff
@@ -36,7 +36,8 @@ trait Prototypes {
     cleanAll := Def.taskDyn {
       val allProjects = ScopeFilter(inAnyProject)
       clean.all(allProjects)
-    }.value
+    }.value,
+    unmanagedClasspath in Runtime += file("ui/dist")
```
Comment (NataliaLKB) [inline project/Prototypes.scala]: Does this mean if it doesn't exist then the app compiles fine, but then fails in runtime? If so, I don't think thats the best behaviour. 

As it stands now if you don't compile(move) the clientside assets to the correct place (running `make compile` or `make watch`) then serverside compilation fails. That makes sense to me and I would rather this behave the same

Comment (NataliaLKB) [inline project/Prototypes.scala]: Unless this will only be temporary to get us started in which case that is fine... But lets add a note to remove.

File: common/app/uiComponent/core/JavascriptRendering.scala [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+trait JavascriptRendering extends Logging {
+
+  def javascriptFile: String
+
+  private implicit val scriptContext = createContext()
+  private val memoizedJs: Try[EvalResult] = loadJavascript()
+
+  def render(props: Option[JsValue] = None)(implicit ac: ApplicationContext): Try[String] = for {
+    propsObject <- encodeProps(props)
+    js <- if(ac.environment.mode == Mode.Dev) loadJavascript() else memoizedJs
```
Comment (NataliaLKB) [inline common/app/uiComponent/core/JavascriptRendering.scala]: I don't remember why we need to do something different for dev then PROD. Is it worth adding a comment to clarify?

Comment (TBonnin) [inline common/app/uiComponent/core/JavascriptRendering.scala]: In prod the javascript bundle doesn't change so we can load it once.
When developing you want every rendering to take into account the lastest change made to the javascript bundle.

I can add a comment to explain it

Comment (NataliaLKB) [inline common/app/uiComponent/core/JavascriptRendering.scala]: Thanks @TBonnin!

File: project/Prototypes.scala (hunk 3 above)
Comment (TBonnin) [inline project/Prototypes.scala]: This line tells sbt to include the`ui/dist` directory into the classpath, so its content is available.

I agree with you that failing at compilation if/when the javascript bundle is not present (or even better when it doesn't compile) is the better solution but I figured we could do without it for now. It is on my list to add asap though.

Can you show me the current behaviour with the client-side assets you are describing in the comment? Thank you

Comment (NataliaLKB) [inline project/Prototypes.scala]: > Can you show me the current behaviour with the client-side assets you are describing in the comment

I can try!

File: common/app/uiComponent/UIComponent.scala [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package uiComponent
+
+import play.api.libs.json.JsValue
+
+trait UIComponent {
+  def asJson: Option[JsValue]
```
Comment (jfsoul) [inline common/app/uiComponent/UIComponent.scala]: My understanding is that `asJson` is the state or props - why not call it that?

File: common/app/uiComponent/core/RenderingActor.scala [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import uiComponent.UIComponent
+import scala.util.Try
+
+case class Rendering(component: UIComponent, ac: ApplicationContext)
+
+case class RenderingException(error: String) extends RuntimeException(error)
+
+class RenderingActor extends Actor with JavascriptRendering {
+
+  override def javascriptFile: String = "ui.bundle.server.js"
```
Comment (jfsoul) [inline common/app/uiComponent/core/RenderingActor.scala]: I'm not that familiar with the client-side part of this - what would `ui.bundle.server.js` be?

File: common/app/uiComponent/core/RenderingActor.scala [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package uiComponent.core
+
+import akka.actor.Actor
+import model.ApplicationContext
+import uiComponent.UIComponent
+import scala.util.Try
+
+case class Rendering(component: UIComponent, ac: ApplicationContext)
```
Comment (jfsoul) [inline common/app/uiComponent/core/RenderingActor.scala]: I think often the available messages for an actor go in a companion object, but in this case I suppose it may be cleaner like this.

File: common/test/uiComponent/core/RenderingActorTest.scala [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import scala.util.{Failure, Success, Try}
+
+@DoNotDiscover class RenderingActorTest
+  extends FlatSpec
+  with ConfiguredTestSuite
+  with WithTestContext
+  with Matchers {
+
+  lazy val actorSystem: ActorSystem = app.actorSystem
+  implicit lazy val timeout = new Timeout(2.seconds)
```
Comment (TBonnin) [inline common/test/uiComponent/core/RenderingActorTest.scala]: the `?` method need an implicit timeout

File: common/app/uiComponent/UIComponent.scala (hunk 5 above)
Comment (TBonnin) [inline common/app/uiComponent/UIComponent.scala]: good question. Maybe state would be more comprehensible

File: common/app/uiComponent/core/RenderingActor.scala (hunk 6 above)
Comment (TBonnin) [inline common/app/uiComponent/core/RenderingActor.scala]: The react app responsible for the rendering :D

File: common/app/uiComponent/TestComponent.scala [hunk 9]
```diff
@@ -3,5 +3,5 @@ package uiComponent
 import play.api.libs.json.JsValue
 
 object TestComponent extends UIComponent {
-  override def asJson: Option[JsValue] = None
+  override def state: Option[JsValue] = None
```
Comment (NataliaLKB) [inline common/app/uiComponent/TestComponent.scala]: Its really `props` isn't it? In react state means something different and in the UI part we are receiving props.

File: common/test/uiComponent/core/RenderingActorTest.scala [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      case Success(s) => s should not be(empty)
+      case Failure(e) => fail("A string should have been returned")
+    }
+  }
+
+  "Sending unknown message" should "return a rendering exception" in {
+    val f = (actor ? "unknown message").mapTo[Try[String]]
+    Await.result(f, timeout.duration) match {
+      case Success(s) => fail("Exception should have been returned")
+      case Failure(e) => e should be(an[RenderingException])
```
Comment (TBonnin) [inline common/test/uiComponent/core/RenderingActorTest.scala]: failAs?

File: common/app/uiComponent/core/RenderingActor.scala (hunk 6 above)
Comment (NataliaLKB) [inline common/app/uiComponent/core/RenderingActor.scala]: There will be a separate PR for that from @sndrs and @SiAdcock

File: common/test/uiComponent/core/RenderingActorTest.scala (hunk 10 above)
Comment (jfsoul) [inline common/test/uiComponent/core/RenderingActorTest.scala]: the `FailAsMatcher` helper you wrote

**Review by NataliaLKB — APPROVED** [review]
This is really cool @TBonnin ! I am sure there are lots of things we would like to improve, but for now this is enough to keep iterating on.

File: common/app/uiComponent/core/RenderingActor.scala (hunk 6 above)
Comment (jfsoul) [inline common/app/uiComponent/core/RenderingActor.scala]: I think I need to see that PR to fully comprehend this one 😆

File: common/test/uiComponent/core/RenderingActorTest.scala (hunk 10 above)
Comment (TBonnin) [inline common/test/uiComponent/core/RenderingActorTest.scala]: :D I even forgot I wrote it. Thx

File: common/app/uiComponent/core/RenderingActor.scala (hunk 6 above)
Comment (NataliaLKB) [inline common/app/uiComponent/core/RenderingActor.scala]: Potentially looking at the tests might help more. Right now the serverside part doesn't require `ui.bundle.server.js` to be any more complicated then https://github.com/guardian/frontend/pull/17515/files#diff-fa7954d48ee0602561496f1794e06217R1

Comment (NataliaLKB) [inline common/app/uiComponent/core/RenderingActor.scala]: > lower level parts

Does that mean smaller components? If not, I don't understand

Comment (NataliaLKB) [inline common/app/uiComponent/core/RenderingActor.scala]: If it does mean smaller components, then its really the naming which needs to change. Right now (subject to change) the `RenderingActor` will only serve a full page. Yes this full page is made up of components, but its completely up to the JS app to figure that all out. 

In other words, the serverside code doesn't "compose" anything at all. It is only aware of the final product. 

Maybe `UIComponent` should be renamed to `pageToRender` or `html` or something more clear.

Comment (jfsoul) [inline common/app/uiComponent/core/RenderingActor.scala]: Let's not worry about it, I'm getting off topic anyway.  Nothing wrong with the PR, just keen to see how the react app comes together with this.

Comment (TBonnin) [inline common/app/uiComponent/core/RenderingActor.scala]: The RenderingActor doesn't know (or care) about what the javascript rendering returns. It might be a full html page, it might be a simple button, it doesn't matter as long as it is a string.
I am happy to rename `UIComponent` with something that encapsulates this is the representation of something that can be rendered, regardless of its complexity.
Writing this make me think that something like `Renderable` could be an acceptable `trait` name :D

Comment (TBonnin) [inline common/app/uiComponent/core/RenderingActor.scala]: `Renderable` it is :D 
thx both for your valuable feedback
